astropy/io/fits/convenience.py

```diff
@@ -494,6 +494,8 @@ def table_to_hdu(table):
         fits_rec = FITS_rec.from_columns(np.array(table.filled()))
         table_hdu = BinTableHDU(fits_rec)
 
+    table_hdu.columns._add_listener(table_hdu)
```

Comment:
This kind of thing should already be done in BinTableHDU or its parent class. Also it works with a `BinTableHDU` created manually:
```
In [1]: %astropy
Numpy 1.11.3
Astropy 2.0.dev17737

In [2]: >>> from astropy.io import fits
   ...: >>> import numpy as np
   ...: >>> a1 = np.array(['NGC1001', 'NGC1002', 'NGC1003'])
   ...: >>> a2 = np.array([11.1, 12.3, 15.2])
   ...: >>> col1 = fits.Column(name='target', format='20A', array=a1)
   ...: >>> col2 = fits.Column(name='V_mag', format='E', array=a2)
   ...: 

In [3]: tbhdu = fits.BinTableHDU.from_columns([col1, col2])

In [5]: tbhdu.header
Out[5]: 
XTENSION= 'BINTABLE'           / binary table extension                         
BITPIX  =                    8 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                   24 / length of dimension 1                          
NAXIS2  =                    3 / length of dimension 2                          
PCOUNT  =                    0 / number of group parameters                     
GCOUNT  =                    1 / number of groups                               
TFIELDS =                    2 / number of table fields                         
TTYPE1  = 'target  '                                                            
TFORM1  = '20A     '                                                            
TTYPE2  = 'V_mag   '                                                            
TFORM2  = 'E       '                                                            

In [6]: tbhdu.columns['target'].unit = 'test'

In [7]: tbhdu.header
Out[7]: 
XTENSION= 'BINTABLE'           / binary table extension                         
BITPIX  =                    8 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                   24 / length of dimension 1                          
NAXIS2  =                    3 / length of dimension 2                          
PCOUNT  =                    0 / number of group parameters                     
GCOUNT  =                    1 / number of groups                               
TFIELDS =                    2 / number of table fields                         
TTYPE1  = 'target  '                                                            
TFORM1  = '20A     '                                                            
TUNIT1  = 'test    '                                                            
TTYPE2  = 'V_mag   '                                                            
TFORM2  = 'E       '    
```